From b1956b274d84f240caf14906a405c83d08def0ea Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:52:15 -0400
Subject: [iommu] amd: fix enabling exclusion range for an exact device

Message-id: <20140905015215.31297.7001.stgit@gir.stowe>
Patchwork-id: 91358
O-Subject: [RHEL7.1 PATCH 65/65] iommu/amd: fix enabling exclusion range for an exact device
Bugzilla: 1129880 1087643
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1087643
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129880
Upstream Status: 2c16c9fdacbbf3bd8ff6390e12c5d28b3f2c78ac

commit 2c16c9fdacbbf3bd8ff6390e12c5d28b3f2c78ac
Author: Su Friendy <friendy.su@sony.com.cn>
Date:   Wed May 7 13:54:52 2014 +0800

    iommu/amd: fix enabling exclusion range for an exact device

    set_device_exclusion_range(u16 devid, struct ivmd_header *m) enables
    exclusion range for ONE device. IOMMU does not translate the access
    to the exclusion range from the device.

    The device is specified by input argument 'devid'. But 'devid' is not
    passed to the actual set function set_dev_entry_bit(), instead
    'm->devid' is passed. 'm->devid' does not specify the exact device
    which needs enable the exclusion range. 'm->devid' represents DeviceID
    field of IVMD, which has different meaning depends on IVMD type.

    The caller init_exclusion_range() sets 'devid' for ONE device. When
    m->type is equal to ACPI_IVMD_TYPE_ALL or ACPI_IVMD_TYPE_RANGE,
    'm->devid' is not equal to 'devid'.

    This patch fixes 'm->devid' to 'devid'.

    Signed-off-by: Su Friendy <friendy.su@sony.com.cn>
    Signed-off-by: Tamori Masahiro <Masahiro.Tamori@jp.sony.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/amd_iommu_init.c b/drivers/iommu/amd_iommu_init.c
index ddd0b1f..62b2333 100644
--- a/drivers/iommu/amd_iommu_init.c
+++ b/drivers/iommu/amd_iommu_init.c
@@ -788,7 +788,7 @@ static void __init set_device_exclusion_range(u16 devid, struct ivmd_header *m)
    * per device. But we can enable the exclusion range per
    * device. This is done here
    */
-  set_dev_entry_bit(m->devid, DEV_ENTRY_EX);
+  set_dev_entry_bit(devid, DEV_ENTRY_EX);
   iommu->exclusion_start = m->range_start;
   iommu->exclusion_length = m->range_length;
  }
-- 
1.7.1